glue/plugins/tools/spectrum_tool/qt/spectrum_tool.py

```diff
@@ -68,50 +68,52 @@ def spectrum(data, attribute, roi, slc, zaxis):
             data = data[attribute, tuple(slc)]   # attribute is Primary,
             finite = np.isfinite(data)
 
-            assert data.ndim == ndim
-
-            for i in reversed(list(range(ndim))):
-                if i != zaxis:  # index of the zaxis in dataset
-                    data = np.nansum(data, axis=i)
-                    finite = finite.sum(axis=i)
-
-            assert data.ndim == 1
-            assert data.size == nz
-
-            data = (1. * data / finite).ravel()
-
         elif isinstance(roi, core_roi.PolygonalROI):
 
             # both circular and polygonal selection will call here
             data = data[attribute]
 
-            slice_shape = []
+            # slice_data_shape is identical with data shape for getting masked data
+            slice_data_shape = []
             for i in range(ndim):
                 if i != zaxis:
-                    slice_shape.append(data.shape[i])
-            slice_data = np.ones(slice_shape)
-            posdata = np.argwhere(slice_data)
-
-            # mask for each slice
-            mask = roi.contains(posdata[:, 0], posdata[:, 1])
-
-            return_data = []
-            for i in range(nz):
-                select_data = np.take(data, i, axis=zaxis).ravel()[mask]
-                finite = np.isfinite(select_data).sum()
-                return_data.append(1. * np.nansum(select_data) / finite)
+                    slice_data_shape.append(data.shape[i])
+
+            # slice_roi_shape is relevant with xyz_axis order for defining roi region
+            slice_roi_shape = np.ones(len(slice_data_shape))
+            slice_roi_shape[0] = data.shape[xaxis]
+            slice_roi_shape[1] = data.shape[yaxis]
+            # for 4th dim
+            for j in range(ndim):
+                if j not in [xaxis, yaxis, zaxis]:
+                    slice_roi_shape[2] = data.shape[j]
+            posdata = np.argwhere(np.ones(slice_roi_shape))
+
+            # mask for each slice and expand to all
+            mask = roi.contains(posdata[:, 0], posdata[:, 1]).reshape(slice_data_shape)
+            mask = np.expand_dims(mask, axis=zaxis).repeat(nz, zaxis)
+
+            # make sure data.ndim == ndim
+            new_shape = np.ones(ndim)
+            new_shape[zaxis] = nz
+            new_shape[xaxis] = -1
+            data = data[mask].reshape(new_shape)  # data[mask].shape is (n, )
+            finite = np.isfinite(data)
 
-            data = np.array(return_data)
+        assert data.ndim == ndim
 
-            assert data.ndim == 1
-            assert data.size == nz
+        for i in reversed(list(range(ndim))):
+            if i != zaxis:  # index of the zaxis in dataset
+                data = np.nansum(data, axis=i)
+                finite = finite.sum(axis=i)
 
-        else:
-
-            raise TypeError("Unexpected ROI type: {0}".format(type(roi)))
```

Comment:
I personally think this kind of Error won't appear so I moved it, coz this line can only be called when the `if ROI.instance` passed, but if later we got similar errors we could add it back ?